Adding ContainerImage subclasses

enoodle opened the pull request:
This will add `ContainerImage` subclasses: `BasicContainerImage` and `OpenshiftContainerImage`. This will allow to treat container images from different sources differently.

Openshift provider complementing PR: https://github.com/ManageIQ/manageiq-providers-openshift/issues/23

Fryguy commented on db/migrate/20170613144708_create_openshift_image.rb:
```diff
@@ -0,0 +1,12 @@
+  def change
+    add_column :container_images, :type, :string
+    ContainerImage.update_all(:type => "OpenshiftContainerImage")
+    ContainerImage.where(:command => nil).update_all(:type => "BasicContainerImage")
```
- These 2 data migration bits cannot be in a `change` method as they are not reversible.  Please split this into an up and a down.
- Please wrap data migrations in a say_with_time like the other data migrations
- Data migrations *must* have a migration spec

enoodle replied:
:+1:

Fryguy commented on db/migrate/20170613144708_create_openshift_image.rb:
```diff
@@ -0,0 +1,12 @@
+    add_column :container_images, :type, :string
+    ContainerImage.update_all(:type => "OpenshiftContainerImage")
+    ContainerImage.where(:command => nil).update_all(:type => "BasicContainerImage")
+    add_column :container_images, :ems_ref, :string
```
Why is this here twice?

enoodle replied:
First column adds `type` and here it adds `ems_ref`. they are different.

Fryguy replied:
Ah thank you...totally misread.

Fryguy commented on app/models/openshift_container_image.rb:
```diff
@@ -0,0 +1,13 @@
+class OpenshiftContainerImage < ContainerImage
```
This seems openshift specific and probably should go into the openshift provider gem itself.

enoodle replied:
:+1:

Fryguy commented on app/models/miq_action.rb:
```diff
@@ -742,7 +742,7 @@ def action_container_image_annotate_deny_execution(action, rec, inputs)
     end
 
-    unless rec.digest.present?
+    unless rec.kind_of?(OpenshiftContainerImage)
```
Since the original code is OpenShift specific already, maybe we can let this go, but it would be preferable if we could avoid provider-specifics.  @simon3z Thoughts?

Fryguy commented on app/models/ems_refresh/save_inventory_container.rb:
```diff
@@ -308,6 +308,7 @@ def save_container_images_inventory(ems, hashes, target = nil)
 
     hashes.each do |h|
       h[:container_image_registry_id] = h[:container_image_registry][:id] unless h[:container_image_registry].nil?
+      h[:type] = h[:ems_ref] ? 'OpenshiftContainerImage' : 'BasicContainerImage'
```
We cannot have provider specifics here.

This can possibly be done differently by setting the :type in the parser itself.  Then here do `h[:type] ||= 'BasicContainerImage'`.  Maybe?  I'm not exactly sure what a BasicContainerImage even is and whether or not it belongs in core or in the OpenShift specific provider code.

enoodle replied:
Ok, will do. `BasicContainerImage` is not an Openshift related class.

Fryguy commented on app/models/container_image.rb:
```diff
@@ -11,6 +11,7 @@ class ContainerImage < ApplicationRecord
   DOCKER_PULLABLE_PREFIX = "docker-pullable://".freeze
   DOCKER_PREFIXES = [DOCKER_IMAGE_PREFIX, DOCKER_PULLABLE_PREFIX].freeze
 
+  belongs_to :parent, :polymorphic => true
```
To me it sounds like what you are trying to get is genealogy, in which case, containers should use the genealogy aspects that we already have for VMs (i.e. we have a relationship tree that show which instance came from which image, and then which image came from another image all the way back)  @simon3z I'm not sure if this is already being done elsewhere in the containers code, but if so, we should make it consistent.

enoodle replied:
I didn't mean to create a genealogy structure here, Just to have all container image types inherit from the same base so that they can be addressed the same and share code.

Fryguy replied:
If so, then `parent` might be a bad name choice as it conflicts with the Relationship code, should that be used in the future.

Fryguy commented on app/models/basic_container_image.rb:
```diff
@@ -0,0 +1,4 @@
+class BasicContainerImage < ContainerImage
```
What is a BasicContainerImage exactly?

enoodle replied:
It is a class to represent a container image not from a specific provider. This is so that it will be easy to know what is the source of the image.

moolitayer replied:
@enoodle why do we need `BasicContainerImage` in addition to `ContainerImage`?

enoodle replied:
It will make it easy to address all the container images that don't have an entity in Openshift/OtherContainerProvider managing them. It will hopefully be handy with the UI.

enoodle commented:
@moolitayer done. PTAL at https://github.com/ManageIQ/manageiq-providers-openshift/issues/23 too

moolitayer commented on app/models/container_image.rb:
```diff
@@ -11,6 +11,7 @@ class ContainerImage < ApplicationRecord
   DOCKER_PULLABLE_PREFIX = "docker-pullable://".freeze
   DOCKER_PREFIXES = [DOCKER_IMAGE_PREFIX, DOCKER_PULLABLE_PREFIX].freeze
 
+  belongs_to :container_image, :polymorphic => true
```
?

enoodle replied:
I though we needed this line, but now I dont think so anymore. I will remove this.

https://github.com/ManageIQ/manageiq/pull/15386#discussion_r122469016

moolitayer commented:
@enoodle after investigation I think we will be better of with only two classes, so that image.type is always `ContainerImage` or `ManageIQ::Providers::Openshift::ContainerManager::ContainerImage`. 
I don't expect it to cause any problems in the UI side.

moolitayer commented on db/migrate/20170613144708_create_openshift_image.rb:
```diff
@@ -0,0 +1,21 @@
+      add_column :container_images, :type, :string
+      ContainerImage.update_all(:type => "ManageIQ::Providers::Openshift::ContainerManager::ContainerImage")
+      ContainerImage.where(:author => nil).update_all(:type => "ContainerImage")
+      add_column :container_images, :ems_ref, :string
```
Do we have a value to update for the Openshift::ContainerManager::ContainerImage ?

moolitayer replied:
If this is only available during refresh, please make sure it will be populated in the next refresh

moolitayer commented on app/models/manageiq/providers/container_manager.rb:
```diff
@@ -13,6 +13,7 @@ class ContainerManager < BaseManager
     has_many :container_limits, :foreign_key => :ems_id, :dependent => :destroy
     has_many :container_image_registries, :foreign_key => :ems_id, :dependent => :destroy
     has_many :container_images, :foreign_key => :ems_id, :dependent => :destroy
+    has_many :basic_container_images, :foreign_key => :ems_id, :dependent => :destroy
```
We don't need this any longer right?

moolitayer commented on spec/factories/basic_container_image.rb:
```diff
@@ -0,0 +1,5 @@
+FactoryGirl.define do
```
We don't need this any longer right?

moolitayer commented on spec/factories/openshift_container_image.rb:
```diff
@@ -0,0 +1,5 @@
+FactoryGirl.define do
```
This should be in manageiq-providers-openshift

moolitayer commented on spec/migrations/20170613144708_create_openshift_image_spec.rb:
```diff
@@ -0,0 +1,29 @@
+require_migration
+
+describe CreateOpenshiftImage do
```
Please update, here should be the tests for `ContainerImage`

moolitayer commented on spec/migrations/20170613144708_create_openshift_image_spec.rb:
```diff
@@ -0,0 +1,29 @@
+
+      migrate
+
+      expect(container_image.reload).to have_attributes(:type => "ContainerImage")
```
Please move this expect before the migrate

moolitayer commented on app/models/miq_action.rb:
```diff
@@ -742,7 +742,7 @@ def action_container_image_annotate_deny_execution(action, rec, inputs)
     end
 
-    unless rec.digest.present?
+    unless rec.kind_of?(ManageIQ::Providers::Openshift::ContainerManager::ContainerImage)
```
Please check presence of the annotate method, since that is what we are going to do here

enoodle commented:
@moolitayer I updated miq_action to look for the existence of the annotate method. I also changed the migration to test for the `size` column because I think it is not controlled by the image creator and will never be nil for images that were collected from /oapi/images.

moolitayer commented on spec/migrations/20170613144708_add_sti_to_container_image_spec.rb:
```diff
@@ -0,0 +1,29 @@
+  migration_context :up do
+    it "up" do
+      container_image = container_image_stub.create!
+      openshift_container_image = container_image_stub.create!(:size => 12345)
```
rubocop wants `12_345`

moolitayer commented:
@Fryguy @simon3z Please review

moolitayer commented:
Thanks @enoodle

Fryguy commented on app/models/miq_action.rb:
```diff
@@ -742,7 +742,7 @@ def action_container_image_annotate_deny_execution(action, rec, inputs)
     end
 
-    unless rec.digest.present?
+    unless red.respond_to?(:annotate_deny_execution)
```
Is `red` a typo?  If so are there no tests that would pick this up?

enoodle replied:
I added a test for that.

Fryguy commented on db/migrate/20170613144708_add_sti_to_container_image.rb:
```diff
@@ -0,0 +1,21 @@
+  end
+
+  def up
+    say_with_time("Adding type and ems_ref columns to ContainerImage and setting type") do
```
The say_with_time should only be around the data migration but...not the add_columns (otherwise you are hiding the logging of the add_column calls).

Fryguy commented on db/migrate/20170613144708_add_sti_to_container_image.rb:
```diff
@@ -0,0 +1,21 @@
+  end
+
+  def down
+    say_with_time("Removing type and ems_ref columns from ContainerImage") do
```
No say_with_time needed on the down since you don't have a data migration.

Fryguy commented on spec/migrations/20170613144708_add_sti_to_container_image_spec.rb:
```diff
@@ -0,0 +1,29 @@
+      expect(container_image_stub.find(container_image.id)).not_to respond_to(:type)
+      expect(container_image_stub.find(openshift_container_image.id)).not_to respond_to(:type)
+    end
+  end
```
No need for a down migration test, since there is no data migration.  Otherwise this test is just testing Rails itself, which isn't necessary.

Fryguy commented:
Database migrations have now been moved to the https://github.com/ManageIQ/manageiq-schema repo.  Please see http://talk.manageiq.org/t/new-split-repo-manageiq-schema/2478 for instructions on how to transfer your database migrations.  If this PR contains *only* migrations, I will leave it open for a short time during the transition, after which I will close this if it has not been moved over.

Fryguy commented:
Merged https://github.com/ManageIQ/manageiq-schema/pull/21.  Please see my note about possible record thrashing, which may require another data migration for the ems_ref: https://github.com/ManageIQ/manageiq-schema/pull/21#issuecomment-312681461

Fryguy commented:
Marking as WIP as this is pending the backend change in ManageIQ/manageiq-providers-openshift#23
